lib/muxer.ex

```diff
@@ -0,0 +1,138 @@
+    stream_format = %Membrane.RemoteStream{type: :packetized, content_format: Ogg}
+
+    header_page =
+      Page.create_first(0)
+      |> Page.append_packet!(Membrane.Ogg.Opus.create_id_header(channels))
+      |> Page.finalize(0)
+
+    comment_page =
+      Page.create_subsequent(header_page)
+      |> Page.append_packet!(Membrane.Ogg.Opus.create_comment_header())
+      |> Page.finalize(0)
+
+    first_audio_data_page = Page.create_subsequent(comment_page)
+
+    buffers = [
+      %Buffer{payload: Page.serialize(header_page)},
+      %Buffer{payload: Page.serialize(comment_page)}
+    ]
+
+    {
+      [stream_format: {:output, stream_format}, buffer: {:output, buffers}],
+      %State{state | current_page: first_audio_data_page}
+    }
+  end
+
+  @impl true
+  def handle_buffer(:input, %Buffer{pts: pts} = buffer, _ctx, state) when not is_nil(pts) do
+    packets_to_encapsulate =
+      if pts > state.total_duration do
+        Membrane.Logger.debug(
+          "Stream discontiunuity of length #{Membrane.Time.as_microseconds(pts - state.total_duration, :round) / 1000}ms, using Packet Loss Concealment"
+        )
+
+        Membrane.Ogg.Opus.create_plc_packets(pts, pts - state.total_duration) ++ [buffer]
+      else
+        [buffer]
+      end
+
+    encapsulate_packets(packets_to_encapsulate, state)
+  end
+
+  @impl true
+  def handle_end_of_stream(:input, _ctx, %State{current_page: current_page} = state) do
+    payload =
+      current_page
+      |> Page.finalize(calculate_granule_position(state.total_duration), true)
+      |> Page.serialize()
+
+    {[buffer: {:output, %Buffer{payload: payload}}, end_of_stream: :output], state}
+  end
+
+  @spec calculate_granule_position(Membrane.Time.t()) :: non_neg_integer()
+  defp calculate_granule_position(duration) do
+    (Membrane.Time.as_seconds(duration, :exact) * @fixed_sample_rate)
+    |> Ratio.trunc()
+  end
+
+  @spec encapsulate_packets([Buffer.t() | Membrane.Ogg.Opus.plc_packet()], State.t(), [Action.t()]) ::
+          {[Action.t()], State.t()}
+  defp encapsulate_packets(packets, state, actions \\ [])
+
+  defp encapsulate_packets([first_packet | rest_packets], state, actions) do
+    {new_actions, state} =
+      case Page.append_packet(state.current_page, first_packet.payload) do
+        {:ok, page} ->
+          {[], %State{state | current_page: page}}
+
+        {:error, :not_enough_space} ->
+          complete_page =
+            state.current_page
+            |> Page.finalize(calculate_granule_position(first_packet.pts))
+
+          new_page =
+            Page.create_subsequent(complete_page)
+            |> Page.append_packet!(first_packet.payload)
+
+          {[buffer: {:output, %Buffer{payload: Page.serialize(complete_page)}}],
+           %State{state | current_page: new_page}}
+      end
+
+    encapsulate_packets(
+      rest_packets,
+      %{state | total_duration: state.total_duration + first_packet.metadata.duration},
+      actions ++ new_actions
```

Comment:
since it seems we only accumulate buffers here, let's make it buffers instead of actions and send them all in a single action